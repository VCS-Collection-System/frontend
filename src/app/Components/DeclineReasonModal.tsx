import * as React from 'react';
import { Button, Form, FormGroup, Modal, ModalVariant, Radio, Stack } from '@patternfly/react-core';
import { CheckCircleIcon } from '@patternfly/react-icons';
import { Controller, useForm } from "react-hook-form";

const DeclineReasonModal = ({...props}) => {
  const { control, handleSubmit } = useForm();

  const declineReasons = [
    "Attachments are not legible",
    "Supporting documentation is missing required information",
    "Supporting documentation is not in an acceptable format",
    "Supporting documentation differs from entered employee or vaccination information",
    "Other (more information to be provided)"
  ]

  const onSubmit = (data) => {
    props.sendReview(false, data["decline-reason"]);
  }

  return (
    <Modal
      isOpen={props.isModalOpen}
      onClose={() => props.setIsModalOpen(false)}
      title="Reason For Decline - Select Most Appropriate"
      variant={ModalVariant.medium}
    >
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          name="decline-reason"
          control={control}
          rules={{ required: true }}
          render={({ field, fieldState }) => (
            <FormGroup
              fieldId="decline-reason"
              isRequired
              validated={fieldState.error ? "error" : "default"}
              helperTextInvalid="* Please select a reason to proceed"
            >
              <Stack hasGutter>
                {declineReasons.map((reason, index) => (
                  <Radio
                    key={"radio-" + index}
                    name={"decline-reason-radio" + index}
                    id={"radio-controlled-" + index}
                    isChecked={field.value === reason}
                    onChange={(_, event) => field.onChange(event.currentTarget.value)}
                    label={reason}
                    value={reason}
                  />
                ))}
              </Stack>
            </FormGroup>
          )}
        />
        <FormGroup fieldId="decline-reason-buttons">
          <Button
            icon={<CheckCircleIcon />}
            isDisabled={props.isSendingReview}
            isLoading={props.isSendingReview}
            key="confirm"
            variant="primary"
            type="submit"
          >
            Confirm
          </Button>
          <Button
            key="cancel"
            onClick={() => props.setIsModalOpen(false)}
            variant="plain"
          >
            Cancel
          </Button>
        </FormGroup>
      </Form>
    </Modal>
  )
}

export default DeclineReasonModal;