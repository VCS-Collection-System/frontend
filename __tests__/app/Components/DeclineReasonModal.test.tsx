import React from 'react';
import DeclineReasonModal from '../../../src/app/Components/DeclineReasonModal';
import { act, fireEvent, render, screen } from '@testing-library/react';

const mockSetIsModalOpen = jest.fn()
const mockSendReview = jest.fn()

describe('<DeclineReasonModal />', ()=> {
    it('should be defined', () => {
        expect(DeclineReasonModal).toBeDefined()
    })

    it("should match existing snapshot", () => {
        expect(render(<DeclineReasonModal isModalOpen={true}/>)).toMatchSnapshot()
    })

    it("should display all decline options", () => {
        render(
            <DeclineReasonModal 
                isModalOpen={true}
                setIsModalOpen={mockSetIsModalOpen}
                sendReview={mockSendReview}
            />
        )
        expect(screen.getByLabelText('Attachments are not legible')).toBeTruthy()
        expect(screen.getByLabelText('Supporting documentation is missing required information')).toBeTruthy()
        expect(screen.getByLabelText('Supporting documentation is not in an acceptable format')).toBeTruthy()
        expect(screen.getByLabelText('Supporting documentation differs from entered employee or vaccination information')).toBeTruthy()
        expect(screen.getByLabelText('Other (more information to be provided)')).toBeTruthy()
    })

    it("should be able to choose an option", async () => {
        render(
            <DeclineReasonModal 
                isModalOpen={true}
                setIsModalOpen={mockSetIsModalOpen}
                sendReview={mockSendReview}
            />
        )
        await act(async () => {
            fireEvent.click(screen.getByRole('radio', {name: /Supporting documentation is not in an acceptable format/i}))
            fireEvent.click(screen.getByRole('button', {name: /Confirm/i}))
        })
        expect(mockSendReview).toBeCalledWith(false, 'Supporting documentation is not in an acceptable format')
    })

    it("should not be able to submit without an option", async () => {
        render(
            <DeclineReasonModal 
                isModalOpen={true}
                setIsModalOpen={mockSetIsModalOpen}
                sendReview={mockSendReview}
            />
        )
        await act(async () => {
            fireEvent.click(screen.getByRole('button', {name: /Confirm/i}))
        })
        expect(screen.getByText('* Please select a reason to proceed')).toBeTruthy()
        expect(mockSendReview).toBeCalledTimes(0)
    })

    it("should be able to close modal with Close header", async () => {
        render(
            <DeclineReasonModal 
                isModalOpen={true}
                setIsModalOpen={mockSetIsModalOpen}
                sendReview={mockSendReview}
            />
        )
        await act(async () => {
            fireEvent.click(screen.getByRole('button', {name: /Close/i}))
        })
        expect(mockSetIsModalOpen).toBeCalledTimes(1)
        expect(mockSetIsModalOpen).toBeCalledWith(false)
    })

    it("should be able to close modal with Cancel button", async () => {
        render(
            <DeclineReasonModal 
                isModalOpen={true}
                setIsModalOpen={mockSetIsModalOpen}
                sendReview={mockSendReview}
            />
        )
        await act(async () => {
            fireEvent.click(screen.getByRole('button', {name: /Cancel/i}))
        })
        expect(mockSetIsModalOpen).toBeCalledTimes(1)
        expect(mockSetIsModalOpen).toBeCalledWith(false)
    })
})