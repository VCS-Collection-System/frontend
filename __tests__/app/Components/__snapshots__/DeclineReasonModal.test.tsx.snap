// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeclineReasonModal /> should match existing snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="pf-c-backdrop__open"
  >
    <div
      aria-hidden="true"
    />
    <div>
      <div
        class="pf-c-backdrop"
      >
        <div
          class="pf-l-bullseye"
        >
          <div
            aria-describedby="pf-modal-part-2"
            aria-labelledby="pf-modal-part-1"
            aria-modal="true"
            class="pf-c-modal-box pf-m-md"
            data-ouia-component-id="OUIA-Generated-Modal-medium-1"
            data-ouia-component-type="PF4/ModalContent"
            data-ouia-safe="true"
            id="pf-modal-part-0"
            role="dialog"
          >
            <button
              aria-disabled="false"
              aria-label="Close"
              class="pf-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 352 512"
                width="1em"
              >
                <path
                  d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                />
              </svg>
            </button>
            <header
              class="pf-c-modal-box__header"
            >
              <h1
                class="pf-c-modal-box__title"
                id="pf-modal-part-1"
              >
                
                <span
                  class="pf-c-modal-box__title-text"
                >
                  Reason For Decline - Select Most Appropriate
                </span>
              </h1>
            </header>
            <div
              class="pf-c-modal-box__body"
              id="pf-modal-part-2"
            >
              <form
                class="pf-c-form"
                novalidate=""
              >
                <div
                  class="pf-c-form__group"
                >
                  <div
                    class="pf-c-form__group-control"
                  >
                    <div
                      class="pf-l-stack pf-m-gutter"
                    >
                      <div
                        class="pf-c-radio"
                      >
                        <input
                          aria-invalid="false"
                          class="pf-c-radio__input"
                          data-ouia-component-id="OUIA-Generated-Radio-1"
                          data-ouia-component-type="PF4/Radio"
                          data-ouia-safe="true"
                          id="radio-controlled-0"
                          name="decline-reason-radio0"
                          type="radio"
                          value="Attachments are not legible"
                        />
                        <label
                          class="pf-c-radio__label"
                          for="radio-controlled-0"
                        >
                          Attachments are not legible
                        </label>
                      </div>
                      <div
                        class="pf-c-radio"
                      >
                        <input
                          aria-invalid="false"
                          class="pf-c-radio__input"
                          data-ouia-component-id="OUIA-Generated-Radio-2"
                          data-ouia-component-type="PF4/Radio"
                          data-ouia-safe="true"
                          id="radio-controlled-1"
                          name="decline-reason-radio1"
                          type="radio"
                          value="Supporting documentation is missing required information"
                        />
                        <label
                          class="pf-c-radio__label"
                          for="radio-controlled-1"
                        >
                          Supporting documentation is missing required information
                        </label>
                      </div>
                      <div
                        class="pf-c-radio"
                      >
                        <input
                          aria-invalid="false"
                          class="pf-c-radio__input"
                          data-ouia-component-id="OUIA-Generated-Radio-3"
                          data-ouia-component-type="PF4/Radio"
                          data-ouia-safe="true"
                          id="radio-controlled-2"
                          name="decline-reason-radio2"
                          type="radio"
                          value="Supporting documentation is not in an acceptable format"
                        />
                        <label
                          class="pf-c-radio__label"
                          for="radio-controlled-2"
                        >
                          Supporting documentation is not in an acceptable format
                        </label>
                      </div>
                      <div
                        class="pf-c-radio"
                      >
                        <input
                          aria-invalid="false"
                          class="pf-c-radio__input"
                          data-ouia-component-id="OUIA-Generated-Radio-4"
                          data-ouia-component-type="PF4/Radio"
                          data-ouia-safe="true"
                          id="radio-controlled-3"
                          name="decline-reason-radio3"
                          type="radio"
                          value="Supporting documentation differs from entered employee or vaccination information"
                        />
                        <label
                          class="pf-c-radio__label"
                          for="radio-controlled-3"
                        >
                          Supporting documentation differs from entered employee or vaccination information
                        </label>
                      </div>
                      <div
                        class="pf-c-radio"
                      >
                        <input
                          aria-invalid="false"
                          class="pf-c-radio__input"
                          data-ouia-component-id="OUIA-Generated-Radio-5"
                          data-ouia-component-type="PF4/Radio"
                          data-ouia-safe="true"
                          id="radio-controlled-4"
                          name="decline-reason-radio4"
                          type="radio"
                          value="Other (more information to be provided)"
                        />
                        <label
                          class="pf-c-radio__label"
                          for="radio-controlled-4"
                        >
                          Other (more information to be provided)
                        </label>
                      </div>
                    </div>
                    
                  </div>
                </div>
                <div
                  class="pf-c-form__group"
                >
                  <div
                    class="pf-c-form__group-control"
                  >
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-primary"
                      data-ouia-component-id="OUIA-Generated-Button-primary-1"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="submit"
                    >
                      <span
                        class="pf-c-button__icon pf-m-start"
                      >
                        <svg
                          aria-hidden="true"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          style="vertical-align: -0.125em;"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                          />
                        </svg>
                      </span>
                      Confirm
                    </button>
                    <button
                      aria-disabled="false"
                      class="pf-c-button pf-m-plain"
                      data-ouia-component-id="OUIA-Generated-Button-plain-2"
                      data-ouia-component-type="PF4/Button"
                      data-ouia-safe="true"
                      type="button"
                    >
                      Cancel
                    </button>
                    
                  </div>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
